# epinano_devel is in-development version of Epinano (https://github.com/enovoa/EpiNano).
# prepare reference transcriptome
samtools faidx transcriptome.fasta
java -jar picard.jar CreateSequenceDictionary R=transcriptome.fasta

# mapping reads to reference transcriptome
minimap2 -ax map-ont transcriptome.fasta mod.fastq | samtools view -hSb -o mod.bam -
minimap2 -ax map-ont transcriptome.fasta unm.fastq | samtools view -hSb -o unm.bam -
samtools sort -@ 4 mod.bam mod.sort && mv mod.sort.bam mod.bam
samtools sort -@ 4 unm.bam unm.sort && mv unm.sort.bam unm.bam
samtools index mod.bam
samtools index unm.bam

# calling variants
samtools view -h -F 4 mod.bam |  java -jar ~/softs/jvarkit/dist/sam2tsv.jar -r transcriptome.fasta    > mod.tsv 2>err
samtools view -h -F 4 unm.bam |  java -jar ~/softs/jvarkit/dist/sam2tsv.jar -r transcriptome.fasta    > unm.tsv 2>err

##1 if input tsv files are not very big
### the following two scripts will compute variants frequencies and extract 5mers
python ../py3/PER_SITE.py3 mod.tsv
python ../py3/PER_SITE.py3 unm.tsv

##2 if input tsv files are very big (e.g., a tsv file continaing more than 10000 reads; this is common in real case)
##2.1 split tsv files on read basis, each small tsv file contains 2000 reads
python ../py3/split_tsv_file.py3 mod.tsv 2000
python ../py3/split_tsv_file.py3 unm.tsv 2000
##2.2 compute variants frequencies at each reference site
for i in *small*tsv;do python per_site_var.frequency.py3 $i ;done
cat mod_small_*tsv.freq | python3 combine_multi_per_site_var_frq.py3 > mod.per_site_var
cat unm_small_*tsv.freq | python3 combine_multi_per_site_var_frq.py3 > unm.per_site_var

# generate 5-mer window of per site variant frequencies
## outputs: mod.per_site_var.5mer.csv and unm.per_site_var.5mer.csv
python slide_per_site_var.py3 mod.per.site.var.csv
python slide_per_site_var.py3 unm.per.site.var.csv

# add current intensity information extracted from event table
## current intensity values will be extracted from event table in basecalled fast5 files generated by albacore basecaller.
## since current intensity values are for 5mers, we need the alignment of read 5-mers against reference sequences

##2 get read 5mers associated current intensity values (only support albacore basecalled fast5 files)
for i in mod*MULTI*fast5;do python fast5ToCurrentIntensity.py3 $i;done # both single- and multi- fast5 are acceptable
for i in unm*MULTI*fast5;do python fast5ToCurrentIntensity.py3 $i;done # both single- and multi- fast5 are acceptable
cat mod*MULTI*current.csv > mod.current.CSV
cat unm*MULTI*current.csv > unm.current.CSV

##3 assign current intensity to read and reference kmers
python assign_current_to_per_read_kmer.py3 mod.current.csv mod.per_read_var.5mer.csv mod.per_site_var.5mer.csv
python assign_current_to_per_read_kmer.py3 unm.current.csv unm.per_read_var.5mer.csv unm.per_site_var.5mer.csv

### updates:
event tables can be extracted from both multi- and single- fast5 files
for i in mod.*.fast5;do echo python fast5ToEventTbl.py3 $i > ${i/fast5/tbl};done # it is recommended to run this parallelly
for i in mod.*.fast5;do echo python fast5ToEventTbl.py3 $i > ${i/fast5/tbl};done
